### PR TITLE
Structured log: replace ngaut/log with logrus

### PR DESCRIPTION
**Pull request by andelf**
Changes:

- vendor: logrus for structured log
- vendor: lumberjack for logrotate
- server/util: additional wrapper, to make log format consistent

review link: https://github.com/pingcap/pd/pull/612/files/18ba74fb8628ad0c7f1b75b70af8724bc3ba8840..b99587ae691568efde81b68572e688bac8827fa5 (without vendor changes)

### CODE REVIEW

File: server/util.go [hunk 1]
```diff
@@ -276,31 +287,110 @@ func (rf *redirectFormatter) Format(pkg string, level capnslog.LogLevel, depth i
...
+// Fire implements logrus.Hook interface
+// https://github.com/sirupsen/logrus/issues/63
+func (hook *contextHook) Fire(entry *log.Entry) error {
+	pc := make([]uintptr, 3, 3)
+	cnt := runtime.Callers(6, pc)
+
+	for i := 0; i < cnt; i++ {
+		fu := runtime.FuncForPC(pc[i] - 1)
+		name := fu.Name()
+		if !strings.Contains(name, "github.com/Sirupsen/logrus") &&
```
Comment (siddontang) [inline server/util.go]: extract the contain check to a function

**Comment by siddontang** [comment]
I think we should move the log to pkg `logutil` package.  PD client may use this later. 

Btw, please add test for it.

**Comment by andelf** [comment]
"move the log to pkg logutil package" is #602 😹 

For that log setting up depends on Config, there will be a circular reference between logutil and server.

😞

**Comment by siddontang** [comment]
You can define a LogConfig for the log, and then embed it into ServerConfig. This can decouple the logic.

File: server/util.go [hunk 2]
```diff
@@ -276,31 +287,110 @@ func (rf *redirectFormatter) Format(pkg string, level capnslog.LogLevel, depth i
...
+
+// Format implements logrus.Formatter
+func (f *textFormatter) Format(entry *log.Entry) ([]byte, error) {
+	var b *bytes.Buffer
+	if entry.Buffer != nil {
+		b = entry.Buffer
+	} else {
+		b = &bytes.Buffer{}
+	}
+	b.WriteString(entry.Time.Format("2006/01/02 15:04:05"))
```
Comment (siddontang) [inline server/util.go]: define a const var for the time format

File: server/util.go [hunk 3]
```diff
@@ -276,31 +287,110 @@ func (rf *redirectFormatter) Format(pkg string, level capnslog.LogLevel, depth i
...
+	if file, ok := entry.Data["file"]; ok {
+		b.WriteString(fmt.Sprintf(" %s:%v:", file, entry.Data["line"]))
```
Comment (siddontang) [inline server/util.go]: seem you can use fmt.Fprintf(b, format, args...) directly to reduce the allocation cost for fmt.Sprintf?

**Comment by andelf** [comment]
LogConfig comes up with `[log]`, may be done with #459 .

File: server/util.go (hunk 1 above)
Comment (andelf) [inline server/util.go]: how about a seperate variable

Comment (siddontang) [inline server/util.go]: you mean defining a table variable for these?

Comment (andelf) [inline server/util.go]: for standalone sub-package, a function is ok.
for mess up with server/util.go, another function may not be acceptable.

never mind 😹

**Comment by andelf** [comment]
Update:

- isolated sub-package pkg/logutil
- logutil.LogConfig embedded in server.Config
- tests

Incompatible Changes:

- pd.toml: `[log]` section

**Comment by andelf** [comment]
review link:
https://github.com/pingcap/pd/pull/612/files/18ba74fb8628ad0c7f1b75b70af8724bc3ba8840..ddf4e8a0401d2bf41931e165c2b0384fde1cb794

File: server/config.go [hunk 4]
```diff
@@ -56,10 +57,8 @@ type Config struct {
 	// Etcd onlys support seoncds TTL, so here is second too.
 	LeaderLease int64 `toml:"lease" json:"lease"`
 
-	// Log level.
-	LogLevel string `toml:"log-level" json:"log-level"`
```
Comment (siddontang) [inline server/config.go]: seem this can cause compatibility

Comment (andelf) [inline server/config.go]: Any other package references pd/server?

Comment (siddontang) [inline server/config.go]: users have to change their config after updating. We should avoid this.

Comment (andelf) [inline server/config.go]: For Ansible deployment log-level is specified via command line argument: https://github.com/pingcap/tidb-ansible/blob/master/roles/pd/templates/run_pd.sh.j2

Shall we add a migration support for backward compatibility?

File: conf/config.toml [hunk 5]
```diff
@@ -15,9 +15,11 @@ initial-cluster = "pd=http://127.0.0.1:2380"
 initial-cluster-state = "new"
 
 lease = 3
-log-level = "info"
 tso-save-interval = "3s"
 
+[log]
+log-level = "info"
```
Comment (siddontang) [inline conf/config.toml]: no need log-file here?

Comment (andelf) [inline conf/config.toml]: previously, log-file is always specified via command line arguments.

File: server/config.go (hunk 4 above)
Comment (siddontang) [inline server/config.go]: Not only for ansible, some users deploy manually. As far as I know,  we have only one user who deploys manually, so we should help him changing the config before updating.

File: pkg/logutil/log.go [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+		strings.Contains(name, "github.com/coreos/pkg/capnslog")
+}
+
+// modifyHook injects file name and line pos into log entry.
+type contextHook struct{}
+
+// Fire implements logrus.Hook interface
+// https://github.com/sirupsen/logrus/issues/63
+func (hook *contextHook) Fire(entry *log.Entry) error {
+	pc := make([]uintptr, 3, 3)
```
Comment (siddontang) [inline pkg/logutil/log.go]: `make([]uinptr, 3)` is enough.

File: pkg/logutil/log.go [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+	case capnslog.TRACE:
+		log.Debugf(logStr)
+	}
+}
+
+// Flush only for implementing Formatter.
+func (rf *redirectFormatter) Flush() {}
+
+// isSKippedPackageName tests wether path name is on log library calling stack.
+func isSkippedPackageName(name string) bool {
```
Comment (siddontang) [inline pkg/logutil/log.go]: add a simple test for this function

**Comment by siddontang** [comment]
Rest LGTM

File: pkg/logutil/log.go (hunk 7 above)
Comment (andelf) [inline pkg/logutil/log.go]: meaningless to add tests for such stupid library function calls.
Or else I'll make it a variable before if condition.

File: server/config.go (hunk 4 above)
Comment (andelf) [inline server/config.go]: The default behavior is not changed. Only matters when a user changes default log level or specifies log file path in pd.toml .

File: pkg/logutil/log.go (hunk 7 above)
Comment (andelf) [inline pkg/logutil/log.go]: btw: coverage already covers this.

File: conf/config.toml (hunk 5 above)
Comment (siddontang) [inline conf/config.toml]: should add it but comment here.

File: pkg/logutil/log.go (hunk 7 above)
Comment (siddontang) [inline pkg/logutil/log.go]: why stupid? if we only cover "github.com/Sirupsen/logrus" but miss "github.com/coreos/pkg/capnslog", how do we know it, at run time?

File: conf/config.toml [hunk 8]
```diff
@@ -18,7 +18,8 @@ lease = 3
 tso-save-interval = "3s"
 
 [log]
-log-level = "info"
+level = "info"
```
Comment (siddontang) [inline conf/config.toml]: do we need to update ansible now?

File: pkg/logutil/log_test.go [hunk 9]
```diff
@@ -23,7 +24,8 @@ import (
 )
 
 const (
-	logPattern = `\d\d\d\d/\d\d/\d\d \d\d:\d\d:\d\d ([\w_%!$@.,+~-]+|\\.)+:\d+: \[(fatal|error|warning|info|debug)\] .*?\n`
+	logPattern   = `\d\d\d\d/\d\d/\d\d \d\d:\d\d:\d\d ([\w_%!$@.,+~-]+|\\.)+:\d+: \[(fatal|error|warning|info|debug)\] .*?\n`
+	thisFilename = "log_test.go"
```
Comment (siddontang) [inline pkg/logutil/log_test.go]: what's "this" mean?

File: conf/config.toml (hunk 8 above)
Comment (andelf) [inline conf/config.toml]: leave it what it is now

File: pkg/logutil/log_test.go (hunk 9 above)
Comment (andelf) [inline pkg/logutil/log_test.go]: current source file

**Comment by siddontang** [comment]
PTAL @disksing

File: pkg/logutil/log_test.go (hunk 9 above)
Comment (siddontang) [inline pkg/logutil/log_test.go]: using runtime.Caller(1) to get the current filename may be better.

Comment (siddontang) [inline pkg/logutil/log_test.go]: or I think you can use this path to check in your test directly.

File: pkg/logutil/log_test.go [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+	c.Assert(err, IsNil)
+	c.Assert(entry, Matches, logPattern)
+	// All capnslog log will be trigered in logutil/log.go
+	c.Assert(strings.Contains(entry, "log.go"), IsTrue)
+
+	log.Warnf("this message comes from logrus")
+	entry, err = s.buf.ReadString('\n')
+	c.Assert(err, IsNil)
+	c.Assert(entry, Matches, logPattern)
+	c.Assert(strings.Contains(entry, thisFilename), IsTrue)
```
Comment (siddontang) [inline pkg/logutil/log_test.go]: can use "log_test.go" here directly.

File: pkg/logutil/log.go [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+func stringToLogLevel(level string) log.Level {
+	switch strings.ToLower(level) {
+	case "fatal":
+		return log.FatalLevel
+	case "error":
+		return log.ErrorLevel
+	case "warn":
+		return log.WarnLevel
+	case "warning":
+		return log.WarnLevel
```
Comment (disksing) [inline pkg/logutil/log.go]: case "warn", "warning":

File: pkg/logutil/log.go [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+	if pkg != "" {
+		pkg = fmt.Sprint(pkg, ": ")
+	}
+
+	logStr := fmt.Sprint(pkg, entries)
+
+	// trim square brackets
+	if len(logStr) > 2 && logStr[0] == '[' {
+		logStr = logStr[1 : len(logStr)-1]
+	}
```
Comment (disksing) [inline pkg/logutil/log.go]: When will it starts with '['? And the last char will always be ']' if first is '['?

File: pkg/logutil/log.go [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+	case capnslog.WARNING:
+		log.Warningf(logStr)
+	case capnslog.NOTICE:
+		log.Infof(logStr)
+	case capnslog.INFO:
+		log.Infof(logStr)
+	case capnslog.DEBUG:
+		log.Debugf(logStr)
+	case capnslog.TRACE:
+		log.Debugf(logStr)
```
Comment (disksing) [inline pkg/logutil/log.go]: case sapnslog.DEBUG, sapnslog.TRACE:

File: pkg/logutil/log.go [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+	b.WriteString(entry.Time.Format(defaultLogTimeFormat))
+	if file, ok := entry.Data["file"]; ok {
+		fmt.Fprintf(b, " %s:%v:", file, entry.Data["line"])
+	}
+	fmt.Fprintf(b, " [%s] %s", entry.Level.String(), entry.Message)
+	for k := range entry.Data {
+		if k != "file" && k != "line" {
+			fmt.Fprintf(b, " %v=%v", k, entry.Data[k])
+		}
+	}
```
Comment (disksing) [inline pkg/logutil/log.go]: how about `for k, v := range entry.Data`?

File: pkg/logutil/log.go (hunk 12 above)
Comment (andelf) [inline pkg/logutil/log.go]: All logs from etcd are wrapped in `[]` pairs. and yes.

File: pkg/logutil/log_test.go (hunk 10 above)
Comment (andelf) [inline pkg/logutil/log_test.go]: OK!

**Comment by andelf** [comment]
+Update: add backward compatibility check + warning

File: pkg/logutil/log.go (hunk 7 above)
Comment (andelf) [inline pkg/logutil/log.go]: log_test.go checks against "log.go"(all capnslog) and "log_test.go".

File: server/config.go [hunk 15]
```diff
@@ -175,6 +182,19 @@ func (c *Config) Parse(arguments []string) error {
 		if err != nil {
 			return errors.Trace(err)
 		}
+
+		// Backward compatibility for toml config
+		fmt.Printf("fuck %+v\n", c)
```
Comment (disksing) [inline server/config.go]: ?

Comment (andelf) [inline server/config.go]: for test .

Comment (andelf) [inline server/config.go]: 😆  fixed

File: pkg/logutil/log.go (hunk 7 above)
Comment (siddontang) [inline pkg/logutil/log.go]: It is nonintuitive. Please test it explicitly.

File: server/config.go [hunk 16]
```diff
@@ -176,6 +182,18 @@ func (c *Config) Parse(arguments []string) error {
 		if err != nil {
 			return errors.Trace(err)
 		}
+
+		// Backward compatibility for toml config
+		if c.LogFileDeprecated != "" && c.Log.Filename == "" {
+			c.Log.Filename = c.LogFileDeprecated
+			msg := fmt.Sprintf("log-file in %s is deprecated, use [log] instead", c.configFile)
+			c.WarningMsgs = append(c.WarningMsgs, msg)
```
Comment (siddontang) [inline server/config.go]: output the warning message here directly

File: pkg/logutil/log.go [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+	return nil
+}
+
+// Levels implements logrus.Hook interface.
+func (hook *contextHook) Levels() []log.Level {
+	return log.AllLevels
+}
+
+func stringToLogLevel(level string) log.Level {
```
Comment (siddontang) [inline pkg/logutil/log.go]: add test to cover all the conditions.

File: server/config.go (hunk 16 above)
Comment (siddontang) [inline server/config.go]: ignore it.

Comment (andelf) [inline server/config.go]: Should be logged after logger is initialized. :)

File: server/config.go [hunk 18]
```diff
@@ -85,6 +88,9 @@ type Config struct {
 	electionMs uint64
 
 	configFile string
+
+	// For all warnings during parsing.
+	WarningMsgs []string
```
Comment (andelf) [inline server/config.go]: public field here

Comment (siddontang) [inline server/config.go]: Got it, but can we use a function to do this? exporting fields seems strange.

File: server/config.go (hunk 16 above)
Comment (disksing) [inline server/config.go]: How about returning warning messages as `[]string` and log them afterwards?

File: server/config.go (hunk 18 above)
Comment (andelf) [inline server/config.go]: GetWarningMsgs() 🤣

Comment (siddontang) [inline server/config.go]: LogWarningMsgs

Comment (andelf) [inline server/config.go]: ugly and export a meaningless function with side effect.

Comment (siddontang) [inline server/config.go]: Exporting WarningMsg is too, what is the WarningMsgs for? what I first see is that there is a WarningMsg config.

File: server/config.go (hunk 16 above)
Comment (siddontang) [inline server/config.go]: returning the warning message seems better.

File: server/config.go (hunk 18 above)
Comment (andelf) [inline server/config.go]: ........

File: server/config.go (hunk 16 above)
Comment (siddontang) [inline server/config.go]: any update?

File: pkg/logutil/log_test.go [hunk 19]
```diff
@@ -41,6 +41,16 @@ func (s *testLogSuite) SetUpSuite(c *C) {
 	s.buf = &bytes.Buffer{}
 }
 
+func (s *testLogSuite) TestStringToLogLevel(c *C) {
+	c.Assert(stringToLogLevel("fatal"), Equals, log.FatalLevel)
```
Comment (siddontang) [inline pkg/logutil/log_test.go]: you can use table-driven test here.

```
let tt = []struct {
    name string
    level Level
} {
    {"ERROR", log.ErrorLevel},
    ...
}

for _, t := range tt {
   c.Assert()
}
```

**Comment by siddontang** [comment]
Rest LGTM

PTAL @disksing @nolouch

**Comment by ngaut** [comment]
LGTM

File: server/config.go (hunk 16 above)
Comment (andelf) [inline server/config.go]: Current implementation is OK for future planning.
`config.Parse()`, `logutil.InitLogger()` may all update warning messages.

**Comment by siddontang** [comment]
LGTM

PTAL @disksing

**Comment by shenli** [comment]
Should we also do this in tidb repo?
 /cc @coocood @tiancaiamao

**Comment by disksing** [comment]
LGTM.

**Comment by andelf** [comment]
I'll fix circleCI status.

**Pull request merged**